main.go

```diff
@@ -974,3 +1043,40 @@ type otelErrorHandler struct {
 func (oh otelErrorHandler) Handle(err error) {
 	level.Error(oh.logger).Log("msg", "opentelemetry", "err", err.Error())
 }
+
+//nolint:lll
+func gRPCServer(cfg *config, tenantHeader string, tenantIDs map[string]string, pm *authentication.ProviderManager, authorizers map[string]rbac.Authorizer, logger log.Logger) (*grpc.Server, error) {
+	director := tracesv1.NewHandler(cfg.traces.writeEndpoint,
+		tracesv1.WithLogger(logger),
+	)
+
+	opts := []grpc.ServerOption{
+		// Note that CustomCodec() is deprecated.  The fix for this isn't calling RegisterCodec() as suggested,
+		// because the codec we need to register is also deprecated.  A better fix, if Google removes
+		// the deprecated type, is https://github.com/mwitkow/grpc-proxy/pull/48
+		grpc.CustomCodec(grpcproxy.Codec()), // nolint: staticcheck
+		grpc.UnknownServiceHandler(grpcproxy.TransparentHandler(director)),
+		// Authorization (tenant)
+		grpc.ChainStreamInterceptor(
+			authentication.WithGRPCTenantHeader(tenantHeader, tenantIDs),
+			authentication.WithGRPCAccessToken(),
+			authentication.WithGRPCTenantInterceptors(pm.GRPCMiddlewares),
+			auth.StreamServerInterceptor(
+				authorization.WithGRPCAuthorizers(authorizers, rbac.Write, "traces", logger)),
+		),
```

Comment:
Agreed that we **need** rate limiting to meet our goal of stopping DDOS and customer abuse.
Agreed about the desirability of metrics.

My thought was that the first GRPC PR should be minimal for ease of reviewing it and to unblock @pavolloffay so that he can test the Jaeger backend with the Observatorium headers.

Reply:
As long as we keep track of it, I'm happy to do it separately. Let's do that to keep this moving then.